internal/cmd/gocdk/launch.go

```diff
@@ -187,6 +209,172 @@ func (local *localLauncher) Launch(ctx context.Context, input *LaunchInput) (*ur
+			return nil, xerrors.Errorf("cloud run launch: %w", err)
+		}
+	}
+
+	// Launch on Cloud Run.
+	var env []*cloudrun.EnvVar
+	for i, inputVar := range input.Env {
+		eqIdx := strings.IndexByte(inputVar, '=')
+		if eqIdx == -1 {
+			return nil, xerrors.Errorf("cloud run launch: bad format for specifier env[%d] (got %q)", i, inputVar)
+		}
+		env = append(env, &cloudrun.EnvVar{
+			Name:  inputVar[:eqIdx],
+			Value: inputVar[eqIdx+1:],
+		})
+	}
+	serviceMeta := &cloudrun.ObjectMeta{
+		Name:      serviceName,
+		Namespace: projectID,
+	}
+	serviceSpec := &cloudrun.ServiceSpec{
+		RunLatest: &cloudrun.ServiceSpecRunLatest{
+			Configuration: &cloudrun.ConfigurationSpec{
+				RevisionTemplate: &cloudrun.RevisionTemplate{
+					Spec: &cloudrun.RevisionSpec{
+						Container: &cloudrun.Container{
+							Image: input.DockerImage,
+							Env:   env,
+							// TODO(light): Add liveness and readiness probes.
+						},
+						ContainerConcurrency: 0, // thread-safe
+					},
+				},
+			},
+		},
+	}
+	locationString := "projects/" + projectID + "/locations/" + location
+	serviceString := locationString + "/services/" + serviceName
+	createCall := crl.client.Projects.Locations.Services.Create(locationString, &cloudrun.Service{
+		ApiVersion: "serving.knative.dev/v1alpha1",
+		Kind:       "Service",
+		Metadata:   serviceMeta,
+		Spec:       serviceSpec,
+	})
+	_, err := createCall.Context(ctx).Do()
+	if apiError := (*googleapi.Error)(nil); xerrors.As(err, &apiError) && apiError.Code == http.StatusConflict {
+		// Already exists, add revision.
+		replaceCall := crl.client.Projects.Locations.Services.ReplaceService(serviceString, &cloudrun.Service{
+			ApiVersion: "serving.knative.dev/v1alpha1",
+			Kind:       "Service",
+			Metadata:   serviceMeta,
+			Spec:       serviceSpec,
+		})
+		// Handle error below in the same way a create call is handled.
+		_, err = replaceCall.Context(ctx).Do()
+	} else if err == nil {
+		// Make publicly accessible.
+		policyCall := crl.client.Projects.Locations.Services.SetIamPolicy(serviceString, &cloudrun.SetIamPolicyRequest{
+			Policy: &cloudrun.Policy{
+				Bindings: []*cloudrun.Binding{
+					{
+						Role:    "roles/run.invoker",
+						Members: []string{"allUsers"},
+					},
+				},
+			},
+		})
+		if _, err := policyCall.Context(ctx).Do(); err != nil {
+			return nil, xerrors.Errorf("cloud run launch: %w")
+		}
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("cloud run launch: %w", err)
+	}
+	crl.logger.Printf("Created revision, waiting for service %s to make changes...", serviceName)
+
+	// Wait for it to become ready.
+	tick := time.NewTicker(5 * time.Second)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+		case <-ctx.Done():
+			return nil, xerrors.Errorf("cloud run launch: wait for ready: %w", ctx.Err())
+		}
+
+		currService, err := crl.client.Projects.Locations.Services.Get("projects/" + projectID + "/locations/" + location + "/services/" + serviceName).Context(ctx).Do()
```

Comment:
Don't we have this as `serviceString` already?

Reply:
Yes. Fixed.